components/calendar/src/islamic.rs

```diff
@@ -582,6 +578,248 @@ impl UmmAlQura {
+        _calendar2: &Self,
+        _largest_unit: DateDurationUnit,
+        _smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        date1.0.until(date2.0, _largest_unit, _smallest_unit)
+    }
+
+    fn debug_name(&self) -> &'static str {
+        "IslamicCivil"
+    }
+
+    fn year(&self, date: &Self::DateInner) -> types::FormattableYear {
+        Self::year_as_islamic(date.0.year)
+    }
+
+    fn month(&self, date: &Self::DateInner) -> types::FormattableMonth {
+        date.0.month()
+    }
+
+    fn day_of_month(&self, date: &Self::DateInner) -> types::DayOfMonth {
+        date.0.day_of_month()
+    }
+
+    fn day_of_year_info(&self, date: &Self::DateInner) -> types::DayOfYearInfo {
+        let prev_year = date.0.year.saturating_sub(1);
+        let next_year = date.0.year.saturating_add(1);
+        types::DayOfYearInfo {
+            day_of_year: date.0.day_of_year(),
+            days_in_year: date.0.days_in_year(),
+            prev_year: Self::year_as_islamic(prev_year),
+            days_in_prev_year: Self::days_in_provided_year(prev_year),
+            next_year: Self::year_as_islamic(next_year),
+        }
+    }
+}
+
+impl IslamicCivil {
+    /// Constructs a new Islamic Civil Calendar
+    pub fn new() -> Self {
+        Self
+    }
+
+    // "Fixed" is a day count representation of calendars staring from Jan 1st of year 1 of the Georgian Calendar.
+    // The fixed date algorithms are from
+    // Dershowitz, Nachum, and Edward M. Reingold. _Calendrical calculations_. Cambridge University Press, 2008.
+    //
+    // Lisp code reference:https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2076
+    fn fixed_from_islamic(i_date: IslamicCivilDateInner) -> RataDie {
+        let year = i64::from(i_date.0.year);
+        let month = i64::from(i_date.0.month);
+        let day = i64::from(i_date.0.day);
+
+        RataDie::new(
+            (FIXED_ISLAMIC_EPOCH_FRIDAY.to_i64_date() - 1)
+                + (year - 1) * 354
+                + div_rem_euclid_f64(3.0 + year as f64 * 11.0, 30.0).0 as i64
+                + 29 * (month - 1)
+                + div_rem_euclid_f64(month as f64, 2.0).0 as i64
+                + day,
+        )
+    }
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2090
+    #[allow(clippy::unwrap_used)]
+    fn islamic_from_fixed(date: RataDie) -> Date<IslamicCivil> {
+        let year = div_rem_euclid_f64(
+            ((date.to_f64_date() - FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()) * 30.0) + 10646.0,
+            10631.0,
+        )
+        .0 as i32;
+        let prior_days = date.to_f64_date()
+            - ((Self::fixed_from_islamic(IslamicCivilDateInner(
+                ArithmeticDate::new_from_lunar_ordinals(year, 1, 1).unwrap(), // Safe unwrap due to hardcoded values,
+            )))
+            .to_f64_date());
+        let month = div_rem_euclid_f64((prior_days * 11.0) + 330.0, 325.0).0 as u8;
```

Comment:
Issue: Not immediately clear that this is within the bounds of a u8.  Please add an explanatory comment.

Reply:
Added a comment explaining why.